crates/std_detect/src/detect/os/linux/riscv.rs

```diff
@@ -46,29 +153,172 @@ pub(crate) fn detect_features() -> cache::Initializer {
     );
     let has_i = bit::test(auxv.hwcap, (b'i' - b'a').into());
     // If future RV128I is supported, implement with `enable_feature` here
-    #[cfg(target_pointer_width = "64")]
+    // Checking target_pointer_width instead of target_arch is incorrect since
+    // there are RV64ILP32* ABIs.
+    #[cfg(target_arch = "riscv64")]
     enable_feature(&mut value, Feature::rv64i, has_i);
-    #[cfg(target_pointer_width = "32")]
+    #[cfg(target_arch = "riscv32")]
     enable_feature(&mut value, Feature::rv32i, has_i);
-    #[cfg(target_pointer_width = "32")]
+    // FIXME: e is not exposed in any of asm/hwcap.h, uapi/asm/hwcap.h, uapi/asm/hwprobe.h
+    #[cfg(target_arch = "riscv32")]
     enable_feature(
         &mut value,
         Feature::rv32e,
         bit::test(auxv.hwcap, (b'e' - b'a').into()),
     );
-    enable_feature(
-        &mut value,
-        Feature::h,
-        bit::test(auxv.hwcap, (b'h' - b'a').into()),
-    );
     enable_feature(
         &mut value,
         Feature::m,
         bit::test(auxv.hwcap, (b'm' - b'a').into()),
     );
-    // FIXME: Auxvec does not show supervisor feature support, but this mode may be useful
-    // to detect when Rust is used to write Linux kernel modules.
-    // These should be more than Auxvec way to detect supervisor features.
+    let has_v = bit::test(auxv.hwcap, (b'v' - b'a').into());
+    enable_features(
+        &mut value,
+        &[
+            Feature::v,
+            Feature::zve32f,
+            Feature::zve32x,
+            Feature::zve64d,
+            Feature::zve64f,
+            Feature::zve64x,
+        ],
+        has_v,
+    );
+
+    let mut out = [
+        riscv_hwprobe {
+            key: RISCV_HWPROBE_KEY_IMA_EXT_0,
+            value: 0,
+        },
+        riscv_hwprobe {
+            key: RISCV_HWPROBE_KEY_CPUPERF_0,
+            value: 0,
+        },
+        riscv_hwprobe {
+            key: RISCV_HWPROBE_KEY_MISALIGNED_VECTOR_PERF,
+            value: 0,
+        },
+    ];
+    if _riscv_hwprobe(&mut out) {
+        let mut enable_feature = |feature, enable| {
+            if enable {
+                value.set(feature as u32);
+            }
+        };
+        if out[0].key != -1 {
+            let ima_ext_0 = out[0].value;
```

Comment:
It seems using this key alone is not robust enough (checking the `IMA` base is necessary).

I opened #1770 for my similar proposal with OS-independent extension implication logic and with fixes including this part.

Reply:
Good catch. Earlier versions of this PR actually included that check, but it was accidentally removed when I changed to use hwcap.
https://github.com/rust-lang/stdarch/commit/3203fd550d499bde03f74f02a810b9d1b3ac2a02#diff-dfab5b675012c4be84a409e8c4f92c8e172e797caf07a23516d1fda390fb9818R152